packages/yarnpkg-fslib/sources/algorithms/watchFile.ts

```diff
@@ -0,0 +1,167 @@
+  if (current !== expected) {
+    throw new Error(`Invalid StatWatcher status: expected '${expected}', got '${current}'`);
+  }
+}
+
+// `bigint` can only be set class-wide, because that's what Node does
+export type ListenerOptions = Omit<Required<WatchFileOptions>, 'bigint'>;
+
+export type CustomStatWatcherOptions = {
+  // BigInt Stats aren't currently implemented in the FS layer, so this is a no-op
+  bigint?: boolean,
+};
+
+export class CustomStatWatcher<P extends Path> extends EventEmitter implements StatWatcher {
+  public readonly fakeFs: FakeFS<P>;
+
+  public readonly path: P;
+
+  public readonly bigint: boolean;
+
+  private status: Status = Status.Ready;
+
+  private changeListeners: Map<WatchFileCallback, NodeJS.Timeout> = new Map();
+
+  private lastStats: Stats;
+
+  constructor(fakeFs: FakeFS<P>, path: P, {bigint = false}: CustomStatWatcherOptions = {}) {
+    super();
+
+    this.fakeFs = fakeFs;
+    this.path = path;
+    this.bigint = bigint;
+
+    this.lastStats = this.stat();
+  }
+
+  start() {
+    assertStatus(this.status, Status.Ready);
+    this.status = Status.Running;
+
+    // Node allows other listeners to be registered up to 3 milliseconds
+    // after the watcher has been started, so that's what we're doing too
+    setTimeout(() => {
+      // Per the Node FS docs:
+      // "When an fs.watchFile operation results in an ENOENT error,
+      // it will invoke the listener once, with all the fields zeroed
+      // (or, for dates, the Unix Epoch)."
+      if (!this.fakeFs.existsSync(this.path)) {
+        this.emit(Event.Change, this.lastStats, this.lastStats);
+      }
+    }, 3);
+  }
+
+  stop() {
+    assertStatus(this.status, Status.Running);
+    this.status = Status.Stopped;
+
+    this.emit(Event.Stop);
+  }
+
+  stat() {
+    try {
+      return this.fakeFs.statSync(this.path);
+    } catch (error) {
+      if (error.code === `ENOENT`) {
+        return statUtils.makeEmptyStats();
+      } else {
+        throw error;
+      }
+    }
+  }
+
+  /**
+   * Creates an interval whose callback compares the current stats with the previous stats and notifies all listeners in case of changes.
+   *
+   * @param opts.persistent Decides whether the interval should be immediately unref-ed.
+   */
+  makeInterval(opts: ListenerOptions) {
+    const interval = setInterval(() => {
+      const currentStats = this.stat();
+      const previousStats = this.lastStats;
+
+      if (statUtils.areStatsEqual(currentStats, previousStats))
+        return;
+
+      this.lastStats = currentStats;
+
+      this.emit(Event.Change, currentStats, previousStats);
+    }, opts.interval);
+
+    return opts.persistent ? interval : interval.unref();
+  }
+
+  /**
+   * Registers a listener and assigns it an interval.
+   */
+  registerChangeListener(listener: WatchFileCallback, opts: ListenerOptions) {
+    this.addListener(Event.Change, listener);
+
+    this.changeListeners.set(listener, this.makeInterval(opts));
```

Comment:
We're going to create many intervals - shouldn't we rather have a single one that calls all the callbacks?

Reply:
The way it's currently implemented is more Node-compliant. Node creates a new interval for each added listener and each interval can be persistent (or unref-ed) and can have a different base time-interval (in `ms`). Each interval does the stat-polling independently and then notifies *all* stored listeners. Example:
```ts
const {watchFile} = require('fs');

watchFile('foo', {interval: 2000}, () => console.log('first'));
watchFile('foo', {interval: 5000}, () => console.log('second'));
```

- After 3 milliseconds (the arbitrary number that I've discovered from testing, more on that will be in the other review thread), Node emits a change event to both listeners.
- After 2 seconds, the first interval does the stat polling and let's assume that it sees changes. It then emits a change event which both listeners will pick up.
- After 2 more seconds, the first interval does the stat polling again, the same thing happens.
- After 1 more second, the second interval does the stat polling and let's say it sees changes. It then notifies both listeners.
- After 1 more second, the first interval does the stat polling. If it sees changes, it notifies both listeners again.